x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -381,7 +491,56 @@ public void onResponse(MultiGetResponse multiGetItemResponses) {
                     }
                 }
             }
-            if (this.remainingSlices.decrementAndGet() == 0) {
+            mayFinish();
+        }
+
+        public void onHostsResponse(SearchResponse searchResponse) {
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            for (SearchHit hit : hits) {
+                HostMetadata host = HostMetadata.fromSource(hit.getSourceAsMap());
+                hostsTable.putIfAbsent(host.hostID, host);
+            }
+            log.debug(hostsWatch::report);
+            log.debug("Have [{}] host metadata items", hostsTable.size());
+
+            calculateCO2AndCosts();
+
+            mayFinish();
+        }
+
+        public void calculateCO2AndCosts() {
+            // Do the CO2 and cost calculation in parallel to waiting for frame metadata.
+            StopWatch watch = new StopWatch("calculateCO2AndCosts");
+            CO2Calculator co2Calculator = new CO2Calculator(instanceTypeService, hostsTable, responseBuilder.requestedDuration);
+            CostCalculator costCalculator = new CostCalculator(
+                instanceTypeService,
+                hostsTable,
+                responseBuilder.requestedDuration,
+                responseBuilder.customCostFactor
+            );
+            Map<String, TraceEvent> events = responseBuilder.stackTraceEvents;
+            List<String> missingStackTraces = new ArrayList<>();
+            for (HostEventCount hec : responseBuilder.hostEventCounts) {
+                TraceEvent event = events.get(hec.stacktraceID);
+                if (event == null) {
+                    // this should never happen
+                    missingStackTraces.add(hec.stacktraceID);
+                    continue;
+                }
+                event.annualCO2Tons += co2Calculator.getAnnualCO2Tons(hec.hostID, hec.count);
+                event.annualCostsUSD += costCalculator.annualCostsUSD(hec.hostID, hec.count);
+            }
+            log.debug(watch::report);
+
+            if (missingStackTraces.isEmpty() == false) {
+                StringBuilder stringBuilder = new StringBuilder();
+                Strings.collectionToDelimitedStringWithLimit(missingStackTraces, ",", "", "", 3, stringBuilder);
```

Comment:
The `appendLimit` parameter (here `3`) defines a limit on the number of *characters*, not list items, so I believe `3` is a bit low. Maybe 80 is a good value?